_posts/2024-01-23-ensuring-code-quality-a-guide-to-dynamic-git-hub-pull-request-gates.markdown

```diff
@@ -0,0 +1,309 @@
+this.
+
+### Advantages
+
+*   Only those workflows get executed that are actually required, because the list of relevant files are declared at the
+    `pull_request` trigger.
+    So, the status check overview of the pull request is limited to the relevant checks and no workflow is run unnecessarily.
+*   The triggers and specific trigger conditions can be configured declaratively.<br>
+    Additional triggers besides the `pull_request` that have to run in any case (even without relevant file changes) can
+    be configured in a clean way. And the files restriction of the `pull_request` trigger are also declared where you 
+    would expect them: in the `pull_request` trigger.
+    <br>
+    <br>
+    Declarative config:
+
+    ```yaml
+    on:
+      workflow_dispatch:
+      schedule:
+        - cron: '0 0 * * 1' # 00:00 on Mondays.
+      pull_request:
+        paths:
+          - package-lock.json
+    ```
+
+    Job config:
+
+    ```yaml
+    on:
+      workflow_dispatch:
+      schedule:
+        - cron: '0 0 * * 1' # 00:00 on Mondays.
+      pull_request:
+
+    jobs:    
+      check-if-relevant:
+        runs-on: ubuntu-latest
+        outputs:
+        is-relevant: ${{ '{{' }} (github.event_name == 'pull_request' && steps.relevant-files-changed.outputs.any_changed == 'true') || github.event_name != 'pull_request' }}
+        steps:
+          # https://github.com/marketplace/actions/changed-files
+          - uses: tj-actions/changed-files@v41.0.1
+            if: github.event_name == 'pull_request'
+            id: relevant-files-changed
+            with:
+              files: package-lock.json
+    ```
+
+### Disadvantages
+
+*   Though this approach would reduce the number of workflow runs, it does run the additional `required-status-check`
+    job whenever the dynamically required workflow needs to be executed.
+    And everytime for the workflow that acts as a fallback.<br>
+    On the other hand, it executes very fast and doesn't cost much in terms of GitHub runner time. 
+*   This approach is a little bit more complex as the fallback workflow has to be kept in mind.
+*   It is not transparent for a GitHub repo admin what status checks are actually executed by just taking a look at the
+    branch protection rules, as `required-status-check` or a similar name will be quite generic.
+
+## Conclusion
+
+GitHub's branch protection rules with required status checks are a powerful tool to ensure quality and security.
+They are only missing the ability to configure dynamic status checks that are only required if relevant changed had been
+committed.
+
+However, we have two workarounds to solve this problem:
+1. Check for the relevant file changes in a job of the workflow itself and execute the actual status check only for 
+   relevant changes.
+2. Add a concluding job to all dynamically required workflows and configure it as required status check.
```

Comment:
Want to add something in the lines of:"as this feature evolves, and GitHub gathers more user feedback around such issues, our hope would be that such workarounds will become obsolete."